### ISSUE TITLE
Using script query on percolator fails to parse after upgrade to 8.6 because of subobjects parsing

### ISSUE THREAD

**Issue by mayya-sharipova** [elastic/elasticsearch#97377]
### Elasticsearch Version

8.6 and above

### Java Version

_bundled_

### OS Version

Unix

### Problem Description

While trying to run a script query on an index with a percolator type field but running into error 
```
        "type": "document_parsing_exception",
        "reason": "[10:39] failed to parse: [10:39] [bool] failed to parse field [filter]"
```
This worked in 7.17.0 but does not work in 8.6 (also on current main )

### Steps to Reproduce

1. Create index with percolator field
```
PUT /debug-index-01
{
    "mappings": {
        "properties": {
            "query": {
                "type": "percolator"
            }
        }
    }
}
```
2. Run POST with script on index
```
POST /debug-index-01/_doc
{
    "query": {
        "bool": {
            "filter": [
                {
                    "script": {
                        "script": {
                            "source": "return params.x ? true : false",
                            "lang": "painless",
                            "params": {
                                "x": true
                            }
                        }
                    }
                }
            ]
        }
    }
}
```

On 7.17 it says creates the document.
On 8.8.2 it gives the response below:
```
{
  "error": {
    "root_cause": [
      {
        "type": "document_parsing_exception",
        "reason": "[10:39] failed to parse: [10:39] [bool] failed to parse field [filter]"
      }
    ],
    "type": "document_parsing_exception",
    "reason": "[10:39] failed to parse: [10:39] [bool] failed to parse field [filter]",
    "caused_by": {
      "type": "x_content_parse_exception",
      "reason": "[10:39] [bool] failed to parse field [filter]",
      "caused_by": {
        "type": "x_content_parse_exception",
        "reason": "[10:39] [script] failed to parse field [params]",
        "caused_by": {
          "type": "unsupported_operation_exception",
          "reason": null
        }
      }
    }
  },
  "status": 400
}
```

**Comment by mayya-sharipova** [comment]
Running the doc request with `error_trace=true`, I can get the below stack trace:


```
 "stack_trace": "org.elasticsearch.index.mapper.DocumentParsingException: [10:39] failed to parse: [10:39] 
 [bool] failed to parse field [filter]\n\tat org.elasticsearch.index.mapper.DocumentParser.wrapInDocumentParsingException(DocumentParser.java:215)\n\tat/org.elasticsearch.index.mapper.DocumentParser.internalParseDocument(DocumentParser.java:122)\n\tat/
 org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:72)
 at/org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:78)
 at/org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:1008)\n\t
 at/org.elasticsearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:956)\n\t
 at/org.elasticsearch.index.shard.IndexShard.applyIndexOperationOnPrimary(IndexShard.java:900)\n\t
 at/org.elasticsearch.action.bulk.TransportShardBulkAction.executeBulkItemRequest(TransportShardBulkAction.java:354)\n\t
 at/org.elasticsearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:219)\n\t
 at/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)\n\t
 at/org.elasticsearch.action.bulk.TransportShardBulkAction.performOnPrimary(TransportShardBulkAction.java:286)\n\t
 at/org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:137)\n\t
 at/org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:74)\n\t
 at/org.elasticsearch.action.support.replication.TransportWriteAction$1.doRun(TransportWriteAction.java:215)\n\t
 at/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)\n\t
 at/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)\n\t
 at/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)\n\t
 at/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)\n\t
 at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)\n\t
 at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)\n\t
 at java.base/java.lang.Thread.run(Thread.java:1623)\n
 
 Caused by: org.elasticsearch.xcontent.XContentParseException: [10:39] [bool] failed to parse field [filter]\n\t 
 at org./org.elasticsearch.xcontent.ObjectParser.throwFailedToParse(ObjectParser.java:616)\n\t
 at org./org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:611)\n\t
 at org./org.elasticsearch.xcontent.ObjectParser.parseArray(ObjectParser.java:604)\n\t
 at org./org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:643)\n\t
 at org./org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315)\n\t
 at org./org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:259)\n\t
 at/org.elasticsearch.index.query.BoolQueryBuilder.fromXContent(BoolQueryBuilder.java:293)\n\t
 at/org.elasticsearch.search.SearchModule.lambda$registerQuery$22(SearchModule.java:1195)\n\t
 at org./org.elasticsearch.xcontent.NamedXContentRegistry.parseNamedObject(NamedXContentRegistry.java:147)\n\t
 at/org.elasticsearch.index.query.AbstractQueryBuilder$1.namedObject(AbstractQueryBuilder.java:383)\n\t
 at/org.elasticsearch.index.query.AbstractQueryBuilder.parseInnerQueryBuilder(AbstractQueryBuilder.java:417)\n\t
 at/org.elasticsearch.index.query.AbstractQueryBuilder.parseTopLevelQuery(AbstractQueryBuilder.java:391)\n\ta
 t org.elasticsearch.percolator@8.10.0-SNAPSHOT/org.elasticsearch.percolator.PercolatorFieldMapper.parseQueryBuilder(PercolatorFieldMapper.java:429)\n\t
 at org.elasticsearch.percolator@8.10.0-SNAPSHOT/org.elasticsearch.percolator.PercolatorFieldMapper.parse(PercolatorFieldMapper.java:409)\n\t
 at/org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:376)\n\t
 at/org.elasticsearch.index.mapper.DocumentParser.parseObject(DocumentParser.java:429)\n\t
 at/org.elasticsearch.index.mapper.DocumentParser.innerParseObject(DocumentParser.java:303)\n\t 
 at/org.elasticsearch.index.mapper.DocumentParser.parseObjectOrNested(DocumentParser.java:263)\n\t
 at/org.elasticsearch.index.mapper.DocumentParser.internalParseDocument(DocumentParser.java:113)\n\t... 19 more\nC
 
 Caused by: org.elasticsearch.xcontent.XContentParseException: [10:39] [script] failed to parse field [params]\n\t
 at org./org.elasticsearch.xcontent.ObjectParser.throwFailedToParse(ObjectParser.java:616)\n\tat org./org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:611)
 at org./org.elasticsearch.xcontent.ObjectParser.parseSub(ObjectParser.java:629)\n\t
 at org./org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:315)\n\t
 at org./org.elasticsearch.xcontent.ObjectParser.parse(ObjectParser.java:259)\n\t
 at org./org.elasticsearch.xcontent.ObjectParser.apply(ObjectParser.java:382)\n\t
 at/org.elasticsearch.script.Script.parse(Script.java:401)\n\t
 at/org.elasticsearch.script.Script.parse(Script.java:292)\n\t
 at/org.elasticsearch.index.query.ScriptQueryBuilder.fromXContent(ScriptQueryBuilder.java:95)\n\t
 at/org.elasticsearch.search.SearchModule.lambda$registerQuery$22(SearchModule.java:1195)\n\t
 at org./org.elasticsearch.xcontent.NamedXContentRegistry.parseNamedObject(NamedXContentRegistry.java:147)\n\t
 at/org.elasticsearch.index.query.AbstractQueryBuilder$1.namedObject(AbstractQueryBuilder.java:383)\n\t
 at/org.elasticsearch.index.query.AbstractQueryBuilder.parseInnerQueryBuilder(AbstractQueryBuilder.java:417)\n\t
 at/org.elasticsearch.index.query.BoolQueryBuilder.lambda$static$7(BoolQueryBuilder.java:278)\n\t
 at org./org.elasticsearch.xcontent.AbstractObjectParser.parseArray(AbstractObjectParser.java:418)\n\t
 at org./org.elasticsearch.xcontent.AbstractObjectParser.lambda$declareObjectArrayOrNull$13(AbstractObjectParser.java:304)\n\t
 at org./org.elasticsearch.xcontent.ObjectParser.lambda$declareField$10(ObjectParser.java:431)\n\t
 at org./org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:609)\n\t... 36 more\n
 
 Caused by: java.lang.UnsupportedOperationException\n\t
 at/org.elasticsearch.index.mapper.DotExpandingXContentParser$WrappingParser.map(DotExpandingXContentParser.java:177)\n\t
 at org./org.elasticsearch.xcontent.FilterXContentParser.map(FilterXContentParser.java:63)\n\t
 at org./org.elasticsearch.xcontent.AbstractObjectParser.lambda$declareField$0(AbstractObjectParser.java:174)\n\t
 at org./org.elasticsearch.xcontent.ObjectParser.lambda$declareField$10(ObjectParser.java:431)\n\t
 at org./org.elasticsearch.xcontent.ObjectParser.parseValue(ObjectParser.java:609)\n\t... 52 more\n"
```

**Comment by mayya-sharipova** [comment]
From the stack trace, we can see that the problem is that `DotExpandingXContentParser doesn't support `map` that is necessary to parse params.

We use `DotExpandingXContentParser ` that is necessary to process [subobjects](https://www.elastic.co/guide/en/elasticsearch/reference/current/subobjects.html), which is enabled by default.  

The work-aroud is to disable subobjects:

```js
{
    "mappings": {
        "subobjects": false,
        "properties": {
            "query": {
                "type": "percolator"
            }
        }
    }
}
```

**Comment by heipei** [comment]
Any update on this? We're running into it as well and don't have the option to disable subobjects since the documents we want to percolate rely on subobjects.

**Comment by javanna** [comment]
@heipei which version are you on?

**Comment by heipei** [comment]
Hello @javanna this is on Elasticsearch 8.10.0 running on Docker. Let me know if you need any additional information!

**Comment by javanna** [comment]
Do you have a recreation of this issue in 8.10? Is it while upgrading or with a new index?

**Comment by heipei** [comment]
Hello @javanna, this is an existing index on an existing cluster. Here is the query (simplified) I'm trying to index into the percolate field, and the error I'm receiving:

```"percolate_query": {
  "script_score": {
    "query": {
      "bool": {
        "filter": [
          {
            "query_string": {
              "query": "foo:bar"
            }
          }
        ]
      }
    },
    "script": {
      "source": "long d = 0; for(int i=0; i<params.intchunks.length; i++) { long u = params.intchunks[i]^doc[\"vectors.hash1.int_\" + i.toString()].value; long uCount = u-((u>>>1)&-5270498306774157605L) -((u>>>2)&-7905747460161236407L); d = d + (((uCount+(uCount>>>3))&8198552921648689607L)%63); } return 2048-d",
      "params": {
        "intchunks": [
          1,
          2
        ]
      }
    }
  }
}
````

Error:
```"error": {
  "type": "document_parsing_exception",
  "reason": "[1:517] failed to parse: [1:517] [script_score] failed to parse field [script]",
  "caused_by": {
    "type": "illegal_argument_exception",
    "reason": "[1:517] [script_score] failed to parse field [script]",
    "caused_by": {
      "type": "illegal_argument_exception",
      "reason": "[1:517] [script] failed to parse field [params]",
      "caused_by": {
        "type": "unsupported_operation_exception",
        "reason": "unsupported_operation_exception: null"
      }
    }
  }
}
```

**Comment by javanna** [comment]
Thanks a lot for the recreation @heipei , the bug is fixed in the upcoming 8.12.0 / 8.11.1 releases.

**Comment by heipei** [comment]
@javanna Thanks Luca, I appreciate it!
